drlhp/pref_interface.py

```diff
@@ -0,0 +1,213 @@
+                    remaining_pairs.value = self.remaining_possible_pairs
+                except IndexError:
+                    if idle_cycles > self.max_idle_cycles:
+                        pref_interface_fake_log("Preference interface has gone idle, exiting", logging.INFO)
+                        return
+                    pref_interface_fake_log("Preference interface ran out of untested segments; waiting", logging.DEBUG)
+                    # If we've tested all possible pairs of segments so far,
+                    # we'll have to wait for more segments
+                    idle_cycles += 1
+                    time.sleep(1.0)
+                    self.recv_segments(seg_pipe)
+            s1, s2 = seg_pair
+
+            pref_interface_fake_log(f"Querying preference for segments {s1.hash} and {s2.hash}", logging.DEBUG)
+
+            if not self.synthetic_prefs:
+                pref = self.ask_user(s1, s2)
+            else:
+                if sum(s1.rewards) > sum(s2.rewards):
+                    pref = (1.0, 0.0)
+                elif sum(s1.rewards) < sum(s2.rewards):
+                    pref = (0.0, 1.0)
+                else:
+                    pref = (0.5, 0.5)
+
+                time.sleep(0.25)
+
+            if pref is not None:
+                # We don't need the rewards from this point on, so just send
+                # the frames
+                #print("PrefInterface sending preference to pref pipe!")
+                pref_pipe.put((s1.frames, s2.frames, pref))
+            # If pref is None, the user answered "incomparable" for the segment
+            # pair. The pair has been marked as tested; we just drop it.
+
+            self.recv_segments(seg_pipe)
+
+    def recv_segments(self, seg_pipe):
+        """
+        Receive segments from `seg_pipe` into circular buffer `segments`.
+        """
+        max_wait_seconds = 0.5
+        start_time = time.time()
+        n_recvd = 0
+        while time.time() - start_time < max_wait_seconds:
+            try:
+                segment = seg_pipe.get(block=True, timeout=max_wait_seconds)
+            except queue.Empty:
+                return
+
+            if len(self.segments) < self.max_segs:
+                self.segments.append(segment)
+            else:
+                self.segments[self.seg_idx] = segment
+                self.seg_idx = (self.seg_idx + 1) % self.max_segs
+            n_recvd += 1
+        easy_tf_log.tflog('segment_idx', self.seg_idx)
+        easy_tf_log.tflog('n_segments_rcvd', n_recvd)
+        easy_tf_log.tflog('n_segments', len(self.segments))
+
+    def sample_seg_pair(self):
+        """
+        Sample a random pair of segments which hasn't yet been tested.
+        """
+        segment_idxs = list(range(len(self.segments)))
+        shuffle(segment_idxs)
+        possible_pairs = combinations(segment_idxs, 2)
+        self.remaining_possible_pairs = len(list(deepcopy(possible_pairs))) - len(self.tested_pairs)
```

Comment:
Since `self.segments` is a circular buffer, after the segments buffer is filled, this number won't be accurate.

(It also won't be accurate if segments with the same hash get added to the buffer, which could happen if the agent takes the same actions leading to duplicate segments.)